solver/src/solver/http_solver.rs

```diff
@@ -202,9 +206,47 @@ impl HttpSolver {
                     },
                     mandatory: false,
                 };
-                (*index, uniswap)
+                (*index, pool_model)
             })
-            .collect()
+            .collect();
+        let balancer_cost = self.balancer_cost(gas_price);
+        let weighted_product_models: HashMap<_, AmmModel> = weighted_product_orders
+            .iter()
+            .map(|(index, amm)| {
+                let reserves = amm
+                    .reserves
+                    .iter()
+                    .map(|(token, state)| {
+                        (
+                            *token,
+                            PoolTokenData {
+                                balance: state.balance,
+                                // Bfp -> BigRational -> String -> BigDecimal
+                                weight: BigDecimal::from_str(
+                                    &BigRational::from(state.weight).to_string(),
+                                )
```

Comment:
I see that `weight` is BigDecimal. Is that correct? I think I started a discussion somewhere that we really only use BigDecimal for postgres and we should avoid using it in other places and use BigRational instead.

Regardless, we probably should not convert by going through string. This is bad for performance (doesn't matter that much) and probably incorrect because it depends on how BigRational's to_string implementation works.
If we do want to keep BigDecimal then let's move the conversion into a separate function that we can test. I suspect there are issues with the fractional part.

Alternatively it might be easier to implement a string conversion on Bfp directly instead of going through BigRational.

Reply:
The HTTP solver expects weights in big decimal notation.

> Alternatively it might be easier to implement a string conversion on Bfp directly instead of going through BigRational.

`Bpf` is a Balancer-specific representation of a big decimal. If the HTTP solver is to be generic WRT how it handles weighted pools, we shouldn't expect a `Bfp` weight.

Reply:
I think the correct phrasing is that the http solver expects a string that is the decimal representation of a rational number. This doesn't mean it has to be a BigDecimal in our code, we just need to serialize it properly. BigDecimal happens to do this but also it has a different inner representation than BigRational so the choice isn't obvious.

> Bpf is a Balancer-specific representation of a big decimal.

What I mean is that it might be easier to have `weight` be a string explicitly and then have a specialized BfP to String function that can be used here. This assumes that it is hard for go from BfP to BigDecimal but on second thought it is not. You can use the same division trick from a couple of lines above to avoid going through BigRational -> String> BigDecimal.

I also want to note that BigDecimal vs BigRational is not just about inner representation and to_string. BigRational actually stores rational numbers in perfect precision. BigDecimal does not. For example the number `1 / 3` can only be approximated by BigDecimal and it does so in the division implementation.
I feel that this is another point in favor of using BigRational to store `weight`. Of course we still need to serialize it as a decimal string which could still be implemented as lossy conversion to BigDecimal and then to_string.

Reply:
> I think the correct phrasing is that the http solver expects a string that is the decimal representation of a rational number. This doesn't mean it has to be a BigDecimal in our code, we just need to serialize it properly.

Makes sense. What I meant was that the conversion here should only be for building the data that we pass on the HTTP solver (and not how we internally represent weights), so in that regard it is acting as a glorified string formatter.

> This assumes that it is hard for go from BfP to BigDecimal but on second thought it is not. You can use the same division trick from a couple of lines above to avoid going through BigRational -> String> BigDecimal.

I agree, I also really dislike `BigRational` -> `String` -> `BigDecimal` (and, I'm not 100 sure it would work, as I was under the impression that `BigRational::to_string` produces `1/10` output, while `BigDecimal::from_str` expects `0.1` for example). I was just trying to answer one part of your comment :stuck_out_tongue:.

> I feel that this is another point in favor of using BigRational to store weight.

For some reason, I thought that we already represented weights as `BigRational`. I also agree that this should be the case.

Reply:
Ok folks. I am glad I included my comment here in the code that we were converting `BigRational -> String / String -> BigDecimal` (note that the numerator and denominator as strings should be fine), but I am still lost for suggestions here as to what we actually want to do. 

We already had these discussions regarding how `Bfp -> BigRational` was the ideal conversion we wanted to have in the balancer Bfp module which resulted in #775. So now that we have `Bfp` weight and a conversion tool making this into a `BigRational`... what do we give to the HTTP solver here?

Reply:
I could do `BigDecimal(rational.numerator().to_string()) / BigDecimal(rational.demominator().to_string())`

Reply:
> (note that the numerator and denominator as strings should be fine)

I do not think this is true:
```rust
println!(
    "{:?}",
    BigDecimal::from_str(&BigRational::new(1.into(), 1000.into()).to_string()),
);
```

Produces:
```
Err(ParseBigInt(ParseBigIntError { kind: InvalidDigit }))
```

Reply:
The suggestion is to get the numerator and denominator separately as strings.

Reply:
The minimal change that I would like to see in this PR is to refactor the BfP -> BigDecimal conversion into a separate, testable function with unit tests.
The exact implementation details aren't that important but it is likely that converting the numerator and denominator separately is the way to go.
The other part of the discussion is whether `weights` should be BigRational instead of BigDecmial. We can do this in a separate issue to not block this PR.

Reply:
> The minimal change that I would like to see in this PR is to refactor the BfP -> BigDecimal conversion into a separate, testable function with unit tests.
> 
> The exact implementation details aren't that important but it is likely that converting the numerator and denominator separately is the way to go.
> 
> The other part of the discussion is whether `weights` should be BigRational instead of BigDecmial. We can do this in a separate issue to not block this PR.

Yea I can do this, but I would like to point out that this is what I had already done in #775 before you suggested not to convert to BigDecimal.

Reply:
Thanks for finding that PR. I remembered that discussion but couldn't quickly find it. What I meant there was to always store BigRational in structs and then use one of those serde attributes like `#[serde(with = "u256_decimal")]` to do the conversion to a decimal string. Within this conversion we could go through BigDecimal so that it is just an implementation detail.

Reply:
> The other part of the discussion is whether weights should be BigRational instead of BigDecmial.

AFAIU, weights are currently `Bfp` but should be `BigRational` (because of the reasons you brought up - and we already use `BigRational` for fees for example). I think the `BigDecimal` conversion is only done for the data being sent to the HTTP solver, so its basically used for string formatting (and could just as well be `String` - although personally I prefer `BigDecimal` to avoid type erasure, although it could just as well be implemented as a `BigRational`with a `serde` formatter).

That being said I do think that weights should be `BigRational` in general and not `Bfp`.

> We can do this in a separate issue to not block this PR.

Agreed :+1:.

> The minimal change that I would like to see in this PR is to refactor the BfP -> BigDecimal conversion into a separate, testable function with unit tests.

:+1:

Reply:
I have made a new commit applying the suggested changes. Still not super happy about the unwrap used in `weight: bfp_to_big_decimal(state.weight).unwrap()` but not unwrapping would mean changing the entire method signature just for this one conversion. Suggestions welcome.

Reply:
Resolved by #813